### DISCUSSION TITLE
Deps pruning with MSVC may result in an incorrect build

### DISCUSSION THREAD

**Discussion by kkpattern** [bazelbuild/bazel#19451] [General]
Hi Bazel Community!

Recently, we found that deps pruning with MSVC may result in incorrect build. When built with MSVC, Bazel uses the `/showIncludes` flag to get the actual included header files from a source file. Then Bazel will prune the dependencies based on the output. This behavior is enabled in this [commit](https://github.com/bazelbuild/bazel/pull/17928). The problem is that parsing the `/showIncludes` output can be tricky sometimes.

The main problem is the `/showIncludes` output may come with different languages and encodings. Bazel already supports [14 languages](https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/rules/cpp/ShowIncludesFilter.java#L60). But all with UTF-8 encoding. If the encoding of the MSVC output is not UTF-8(for example, the default encoding for zh-CN locale is cp936), then Bazel will fail to detect the `/showIncludes` output. But this will not result in any error. Bazel will simply think this source file includes no header files. If any of these header files is modified without updating the source file, this source file will not be re-compiled, resulting in an incorrect build. We noticed this when we upgraded Bazel from 5.4.1 to 6.3.2 recently.

One way to fix this is by adding the `VSLANG=1033` environment variable to force MSVC to generate output in English. @meteorcloudy  suggested we can add this env var as a default to the parse_showincludes feature. We tested it, and it works great. Sadly, if we didn't install the English language pack, then MSVC would fall back to one of the languages installed, so `VSLANG=1033` may cause a non-reproducible build and the problem is not entirely solved.

We can fix this problem by installing the English language or changing our build environment to UTF-8 encoding. But I think it's important to avoid Bazel causing incorrect builds without the user's knowledge. Here are some options I came up with:

1. Try to detect the encoding of the `/showIncludes` output. If it's not UTF-8, fail the build. But I don't know if we can find a reliable way to detect the encoding.
2. Make deps pruning with MSVC opt-in. We can disable the parse_showincludes feature by default. In order to do this, we need to add another feature to tell Bazel that MSVC won't produce a dotD file. Currently, Bazel uses the parse_showincludes feature to determine whether the compiler will generate dotD files. If we turn off the parse_showincludes feature, the build will fail because no dotD file is generated.
3. Add a message about this on the [Using Bazel on Windows page](https://bazel.build/configure/windows)

This problem should only affect a few users at present, but it invalidates the most important feature of Bazel: always correct. So I think it is worth solving, and I look forward to your opinions.

Thanks

**Comment by meteorcloudy** [comment]
Thanks for the nice summary! (Context https://github.com/bazelbuild/bazel/pull/17928#issuecomment-1711014182)

> If it's not UTF-8, fail the build. But I don't know if we can find a reliable way to detect the encoding.

Yeah, this seems to be hard.

> We can disable the parse_showincludes feature by default.

I believe header detection is also an important feature to ensure correctness for C++ builds, so disabling it by default doesn't sound ideal.

> Add a message about this on the [Using Bazel on Windows page](https://bazel.build/configure/windows)

Yeah, personally I prefer setting `VSLANG=1033` for `parse_showincludes` feature + documentation. Is the English language pack installed by default or it depends on the user's host config?

**Reply by kkpattern** [reply]
> Is the English language pack installed by default or it depends on the user's host config?

I tested the newest installer with VS community 2022. It depends on the user's host config.

**Reply by meteorcloudy** [reply]
Thanks! Then better documentation is definitely needed for `parse_showincludes` to work.

**Reply by tjgq** [reply]
> If it's not UTF-8, fail the build. But I don't know if we can find a reliable way to detect the encoding.

Do the UTF-8 files generated by MSVC include a [byte order mark](https://en.wikipedia.org/wiki/Byte_order_mark), by any chance? That would be reliable.

**Reply by fmeum** [reply]
> Do the UTF-8 files generated by MSVC include a [byte order mark](https://en.wikipedia.org/wiki/Byte_order_mark), by any chance? That would be reliable.

Just found https://learn.microsoft.com/en-us/cpp/build/reference/utf-8-set-source-and-executable-character-sets-to-utf-8?view=msvc-170&viewFallbackFrom=vs-2019, maybe that has an effect here and could turn a silent into a visible failure?

**Reply by tjgq** [reply]
If the previous suggestions don't work, here's another idea to do the detection reliably: have the C++ repository rule run a simple action that causes the compiler to emit a predictable string containing non-ASCII characters (e.g. a source file with an #error), then check whether the output contains the UTF-8 encoding of the predictable string.

**Reply by kkpattern** [reply]
> have the C++ repository rule run a simple action that causes the compiler to emit a predictable string containing non-ASCII characters (e.g. a source file with an #error), then check whether the output contains the UTF-8 encoding of the predictable string.

I just found out this is how mozilla solved this problem. They create and compile a dummy C source file before build. They know the stdio.h file is included. They extract the prefix according to the output line that contains the stdio.h. Here is the [diff](https://bugzilla.mozilla.org/attachment.cgi?id=472141&action=diff#a/configure.in_sec2).

But Bazel is a more complicated build system. It may be more diffcult for Bazel to do the simliar thing.

**Reply by kkpattern** [reply]
> Just found https://learn.microsoft.com/en-us/cpp/build/reference/utf-8-set-source-and-executable-character-sets-to-utf-8?view=msvc-170&viewFallbackFrom=vs-2019, maybe that has an effect here and could turn a silent into a visible failure?

We tested the `/utf-8` flag, it doesn't effect the output encoding of the cl.exe.

**Reply by fmeum** [reply]
The Mozilla approach looks very promising. I can offer support if you want to turn this into a repository rule.

**Reply by kkpattern** [reply]
Maybe we can implement this in the starlark version of rules_cc? And before the fully migration, we can add a message in the [Using Bazel on Windows page](https://bazel.build/configure/windows) page to tell people how to workaround this problem.

**Reply by fmeum** [reply]
I think that we should fix this right in the Windows toolchain configuration used by default, which (currently) means in `@bazel_tools`. `rules_cc`'s toolchain definitions aren't up-to-date anymore.

I found that CMake also [extracts the prefix of the line](https://github.com/Kitware/CMake/blob/1345928f9693a038153c43f09128bac0419d91ed/Modules/CMakeDetermineCompilerId.cmake#L1150). @kkpattern Do you know which encoding cl uses for the paths it emits? Matching the prefix in any kind of encoding isn't helpful if the path that follows it would use an encoding that would make Bazel unable to match them later.

Ideally we would be able to force the codepage but not the language. As long as the prefix and the file paths are in UTF-8, we can just look for all of them or auto-detect the prefix.

**Reply by fmeum** [reply]
@meteorcloudy Do you know whether there is a way to force a subprocess to run with cp 65001?

**Reply by kkpattern** [reply]
> Do you know which encoding cl uses for the paths it emits?

I'm not sure what encoding is used for the paths because none of our source file paths contain non-ASCII characters. Generally speaking, it should be quite rare to use non-ASCII characters in C/C++ header filename I think? CMake seems utilized this, they match the `/showIncludes` output mainly relies on the two `:` and the following C/C++ header file path. Maybe we can use a similar regex pattern in Bazel?

**Reply by fmeum** [reply]
Having non-ASCII chars in C/C++ header filesnames would probably be rare, but the `/showIncludes` paths are absolute, which probably make them more likely to contain special characters. But I could see that being difficult already with Bazel or really any build system.

> Maybe we can use a similar regex pattern in Bazel?

While we don't have regexes in Starlark, we could do something similar. However, we would need a way to store this information in the `cc_toolchain` for `ShowIncludeFilter` to pick it up.

**Reply by meteorcloudy** [reply]
> Do you know whether there is a way to force a subprocess to run with cp 65001?

I have no idea...

**Reply by kkpattern** [reply]
> Do you know whether there is a way to force a subprocess to run with cp 65001?

There is a Windows API named [GetACP](https://learn.microsoft.com/en-us/windows/win32/api/winnls/nf-winnls-getacp) seems can get the code page of the current process. Maybe we can use it to get the code page then decode and re-encode the `/showIncludes` output into UTF-8. But I'm not familiar with this API, not sure if it's reliable. I only know that CPython is [using it](https://github.com/python/cpython/blob/main/Python/fileutils.c#L910).
